code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/integration/azure/AzureIdentityProvider.java

```diff
@@ -0,0 +1,202 @@
+            case AzureConstants.GROUP_SCOPE:
+                return searchGroups(name, exactMatch);
+            default:
+                return new ArrayList<>();
+        }
+    }
+
+    @Override
+    public Set<Identity> getIdentities(Account account) {
+        if (!isConfigured() || !AzureConstants.CONFIG.equalsIgnoreCase(SecurityConstants.AUTH_PROVIDER.get()) ||
+                !AzureConstants.USER_SCOPE.equalsIgnoreCase(account.getExternalIdType())) {
+            return new HashSet<>();
+        }
+        String accessToken = (String) DataAccessor.fields(account).withKey(AzureConstants.AZURE_ACCESS_TOKEN).get();
+        String refreshToken = (String) DataAccessor.fields(account).withKey(AzureConstants.AZURE_REFRESH_TOKEN).get();
+        if (azureTokenUtils.findAndSetJWT()){
+            ApiRequest request = ApiContext.getContext().getApiRequest();
+            request.setAttribute(AzureConstants.AZURE_ACCESS_TOKEN, accessToken);
+            request.setAttribute(AzureConstants.AZURE_REFRESH_TOKEN, refreshToken);
+            return azureTokenUtils.getIdentities();
+        }
+        String jwt = null;
+        if (!StringUtils.isBlank(accessToken) && SecurityConstants.SECURITY.get()) {
+            AuthToken authToken = authTokenDao.getTokenByAccountId(account.getId());
+            if (authToken == null) {
+                try {
+                    jwt = ProjectConstants.AUTH_TYPE + azureTokenCreator.getAzureToken(accessToken).getJwt();
+                    authToken = authTokenDao.createToken(jwt, AzureConstants.CONFIG, account.getId());
+                    jwt = authToken.getKey();
+                } catch (ClientVisibleException e) {
+                    if (e.getCode().equalsIgnoreCase(AzureConstants.AZURE_ERROR) &&
+                            !e.getDetail().contains("401")) {
+                        throw new ClientVisibleException(ResponseCodes.INTERNAL_SERVER_ERROR,
+                                AzureConstants.JWT_CREATION_FAILED, "", null);
+                    }
+                }
+            } else {
+                jwt = authToken.getKey();
+            }
+
+        }
+        if (StringUtils.isBlank(jwt)){
+            return Collections.emptySet();
+        }
+        ApiRequest request = ApiContext.getContext().getApiRequest();
+        request.setAttribute(AzureConstants.AZURE_JWT, jwt);
+        request.setAttribute(AzureConstants.AZURE_ACCESS_TOKEN, accessToken);
+        request.setAttribute(AzureConstants.AZURE_REFRESH_TOKEN, refreshToken);
+        return azureTokenUtils.getIdentities();
+    }
+
+    private List<Identity> searchGroups(String groupName, boolean exactMatch) {
+        List<Identity> identities = new ArrayList<>();
+
+        AzureUserInfo group;
+        try {
+            group =  azureClient.getAzureGroupByName(groupName);
+            if (group == null){
+                return identities;
+            }
+        } catch (ClientVisibleException e) {
```

Comment:
should we log the exception for debugging purposes here? How do we handle it in the other providers?